### PR TITLE
wx update and cmake

### PR DESCRIPTION
**Pull request by ghost**
This is progress for updating wx and I also have a cmake. It doesn't build yet but I've got it down to some linker errors.

Not ready yet.

### CODE REVIEW

**Comment by ghost** [comment]
I've upgraded to the latest version of [CMarkup](http://www.firstobject.com/dn_markup.htm) for now until I or someone else makes the switch to tinyxml.

The Markup.cpp and Markup.h files should be placed in the code/Torsion directory before cmake configuration and generation.

File: code/Torsion/PreCompiler.cpp [hunk 1]
```diff
@@ -411,10 +411,10 @@ void PreCompilerThread::UpdateOutput( const wxString& path, wxFileOffset& lastRe
       return;
 
    wxString buffer;
-   wxChar* log = buffer.GetWriteBuf( size );
+   wxChar* log = new wxChar(buffer.GetChar(size));
```
Comment (dottools) [inline code/Torsion/PreCompiler.cpp:414]: This is wrong. [wxChar is a typedef depending on unicode setting](http://docs.wxwidgets.org/trunk/group__group__funcmacro__string.html#gad42f64d8c82f1ce4ae58773a89b2d6a7).

File: code/Torsion/shNotebook.h [hunk 2]
```diff
@@ -15,10 +15,10 @@ enum
     shNB_HITTEST_ONICON  = 2,   // on icon
     shNB_HITTEST_ONLABEL = 4,   // on label
     shNB_HITTEST_ONBUTTON = 8,   // 0 = close, 1 = right, 2 = left
-    shNB_HITTEST_ONITEM  = wxNB_HITTEST_ONICON | wxNB_HITTEST_ONLABEL
+    shNB_HITTEST_ONITEM  = 2 | 4
```
Comment (dottools) [inline code/Torsion/shNotebook.h:18]: Again wrong, fix the #include directives if the enums are undefined.

File: code/Torsion/PreCompiler.cpp (hunk 1 above)
Comment (ghost) [inline code/Torsion/PreCompiler.cpp:414]: I'm aware that a lot of the stuff is wrong. Don't worry I'll get to them.

**Comment by dottools** [comment]
There's so many things wrong with this PR I'm not going to bother commenting each and every little thing. So many operations are commented out entirely instead of fixed, and so on. PRs should only be done when they're actually completed. Now don't get me wrong there's a huge amount of great work on getting it compliant with the latest wxWidgets version. However it's still very much an incomplete mess.

**Comment by ghost** [comment]
Who said PRs are only for completed works? Is that in the github bible? This PR is tracking my branch. That's what it's for. I'll rebase later.

**Comment by tomspilman** [comment]
Hi @blackwc.

I think the wxWidgets port work is great...  but i will not merge CMake or CMarkup.

I want to have VS solutions and projects in this repo and not a CMake build script.  When I discussed build scripting in https://github.com/SickheadGames/Torsion/issues/4 it was simply about the glue between the steps and not project generation.  It was mean for replacing this:

 https://github.com/SickheadGames/Torsion/blob/master/msbuild.proj

Not for automating project generation.

As For CMarkup I omitted it from the repo on purpose.  The license in the zip clearly stipulates it is evaluation purposes only.  We're not evaluating it...  I did that 10 years ago... it is great and i'd love to use it, but it is not open source licensed.   I wrote https://github.com/SickheadGames/Torsion/issues/1 specifically for these reasons.  That said if you want to reach out to the CMarkup developers and ask about using it in open source that would be great.

Keeping a WIP PR like this is fine... it doesn't bother anyone and lets us see progress.  Just in the end lets rebase and flatten some of the commits.

**Comment by ghost** [comment]
Understood. I'll look into it.

File: code/Torsion/AutoCompText.cpp [hunk 3]
```diff
@@ -66,8 +66,9 @@ void AutoCompText::InsertRange( size_t pos, const wxString& text )
 
 wxChar* AutoCompText::GetWriteBuf( size_t len ) 
 {
-   wxChar* result = m_Data.GetWriteBuf( len + 1 ); // never allow zero!
-   m_Data.UngetWriteBuf( len );
+
+   wxChar* result = new wxChar(m_Data.GetChar( len + 1 )); // never allow zero!
+   //m_Data.UngetWriteBuf( len );
```
Comment (tomspilman) [inline code/Torsion/AutoCompText.cpp:71]: Note this change is wrong.... it does something completely illogical and will not work.

Note that if memory serves me Torsion was not written for Unicode or wide character support.  If you are trying to port to the latest wxWidgets and at the same time upgrade to wchar or unicode you are just making things more difficult.

Stick to wxWigdets with `wxUSE_UNICODE` set to zero to avoid all this for now.

Comment (ghost) [inline code/Torsion/AutoCompText.cpp:71]: I'm aware of the senseless lines. I just wanted to put torsion on legs, however shaky, for the time being. I am going to attempt to move to unicode as well, yes (maybe, we'll see).

File: code/Torsion/BreakpointPropertyDlg.cpp [hunk 4]
```diff
@@ -137,7 +137,7 @@ void BreakpointPropertyDlg::CreateControls()
     itemBoxSizer4->Add(itemBoxSizer5, 0, wxGROW|wxBOTTOM, 5);
 
     wxStaticText* itemStaticText6 = new wxStaticText( itemDialog1, wxID_STATIC, _("&File:"), wxDefaultPosition, wxSize(30, -1), 0 );
-    itemBoxSizer5->Add(itemStaticText6, 0, wxALIGN_CENTER_VERTICAL|wxALL|wxADJUST_MINSIZE, 0);
+    itemBoxSizer5->Add(itemStaticText6, 0, wxALIGN_CENTER_VERTICAL|wxALL, 0);
```
Comment (tomspilman) [inline code/Torsion/BreakpointPropertyDlg.cpp:140]: FYI....  https://github.com/SickheadGames/Torsion/issues/13

File: code/Torsion/ConfigDlg.cpp [hunk 5]
```diff
@@ -306,7 +306,8 @@ void ConfigDlg::OnOkClick( wxCommandEvent& event )
    m_Config->SetInjectDebugger( m_PropGrid->GetPropertyValueAsBool( "Debugging.OneClick Debugging" ) );
    m_Config->SetExports( m_PropGrid->GetPropertyValueAsBool( "ScriptSense.Build Exports" ) );
 
-   wxDialog::OnOK( event );
+   //wxDialog::OnOK( event );
+   EndModal(wxID_CANCEL);
```
Comment (tomspilman) [inline code/Torsion/ConfigDlg.cpp:310]: Should be `wxID_OK`.

Comment (ghost) [inline code/Torsion/ConfigDlg.cpp:310]: Thanks.

File: code/Torsion/FindReplaceDlg.cpp [hunk 6]
```diff
@@ -260,7 +261,7 @@ void FindReplaceDlg::OnFindClick( wxCommandEvent& event )
                      ( m_WholeWordCtrl->GetValue() ? wxFR_WHOLEWORD : 0 ) );
 
    wxASSERT( GetParent() );
-   GetParent()->ProcessEvent( fevent );
+   reinterpret_cast<MainFrame*>(GetParent())->ProcessEvent(fevent);
```
Comment (tomspilman) [inline code/Torsion/FindReplaceDlg.cpp:264]: I suspect what we want here is:

```
GetParent()->ProcessWindowEvent(fevent);
```

Which should do what `ProcessEvent` did before it was made protected which is propagate the event to main window and any child windows that want to handle it.

Comment (ghost) [inline code/Torsion/FindReplaceDlg.cpp:264]: Right.

File: code/Torsion/GotoDefDialog.cpp [hunk 7]
```diff
@@ -238,7 +238,7 @@ wxIcon GotoDefDialog::GetIconResource( const wxString& name )
 void GotoDefDialog::OnSymbolDoubleClicked( wxListEvent& event )
 {
    m_Selected = event.GetIndex();
-   wxDialog::OnOK( wxCommandEvent() );
+   //wxDialog::OnOK( wxCommandEvent() );
```
Comment (tomspilman) [inline code/Torsion/GotoDefDialog.cpp:241]: This needs to be the `EndDialog(wxID_OK)` in any place where `wxDialog::OnOK(xxx)` was used.

Comment (ghost) [inline code/Torsion/GotoDefDialog.cpp:241]: I appreciate the help and all but this was very rough-rough draft I just threw together. I'm ironing out the creases now. I just wanted to make it compile. I don't want you to think you have to point out every lazy hack I did just to get something to compile.

File: code/Torsion/LexTCS.cxx [hunk 8]
```diff
@@ -0,0 +1,557 @@
+//
```
Comment (tomspilman) [inline code/Torsion/LexTCS.cxx:1]: Is this a copy of the lexer from...

 https://github.com/SickheadGames/Torsion/blob/master/code/wxWidgets/contrib/src/stc/scintilla/src/LexTCS.cxx

?

Note i wrote this lexer, so it is fine to replace the copyright header with the standard Torsion one.

File: code/Torsion/MainFrame.cpp [hunk 9]
```diff
@@ -539,7 +539,7 @@ bool MainFrame::Create( DocManager* manager, wxFrame* frame, const wxString& tit
    m_ProjectSash->SetDefaultBorderSize( 4 );
    m_ProjectSash->SetExtraBorderSize( 0 );
    m_ProjectSash->SetSashVisible( wxSASH_RIGHT, true );
-   m_ProjectSash->SetSashBorder( wxSASH_RIGHT, true ); 
+   //m_ProjectSash->SetSashBorder( wxSASH_RIGHT, true );
```
Comment (tomspilman) [inline code/Torsion/MainFrame.cpp:542]: A note mostly for me....    `SetSashBorder` was depreciated in 2.8 and supposedly isn't used within the code base anymore.

http://blog.gmane.org/gmane.comp.python.wxpython.devel/month=20080401

File: code/Torsion/MainFrame.cpp [hunk 10]
```diff
@@ -2092,14 +2092,16 @@ bool MainFrame::ProcessEvent(wxEvent& event)
    // or output control has focus.
    if ( event.GetId() == wxID_COPY )
    {
-      if (  GetOutputPanel() && 
-            GetOutputPanel()->m_Output == wxWindow::FindFocus() &&
-            GetOutputPanel()->m_Output->ProcessEvent( event ) )
+	   if (GetOutputPanel() &&
+		   GetOutputPanel()->m_Output == wxWindow::FindFocus() &&
+		   //TODO ? GetOutputPanel()->m_Output->ProcessEvent( event ) )
```
Comment (tomspilman) [inline code/Torsion/MainFrame.cpp:2097]: Likely same as i noted above... `ProcessWindowEvent()`

File: code/Torsion/MainFrame.cpp [hunk 11]
```diff
@@ -718,7 +718,7 @@ void MainFrame::OnCloseWindow( wxCloseEvent& event )
    tsGetAutoComp()->SetEnable( false );
 
    // Now let the base clear the rest of the docs.
-   wxDocMDIParentFrame::OnCloseWindow( event );
+   //TODO ? wxDocMDIParentFrame::OnCloseWindow( event );
```
Comment (tomspilman) [inline code/Torsion/MainFrame.cpp:721]: I think you need to do `event.Skip()` here as it allows the event to continue to be processed by the next handler to perform the close.  See http://docs.wxwidgets.org/trunk/classwx_close_event.html

File: code/Torsion/OutputCtrl.cpp [hunk 12]
```diff
@@ -124,8 +124,8 @@ void OutputCtrl::UpdatePrefs()
    StyleSetBackground( wxSTC_STYLE_DEFAULT, tsGetPrefs().GetBgColor() );
    SetCaretForeground( tsGetPrefs().GetDefaultColor() );
 
-	StyleSetForeground( wxSTC_TCS_DEFAULT, tsGetPrefs().GetDefaultColor() );
-	StyleSetBackground( wxSTC_TCS_DEFAULT, tsGetPrefs().GetBgColor() );
+	StyleSetForeground( wxSTC_TCL_DEFAULT, tsGetPrefs().GetDefaultColor() );
+	StyleSetBackground( wxSTC_TCL_DEFAULT, tsGetPrefs().GetBgColor() );
```
Comment (tomspilman) [inline code/Torsion/OutputCtrl.cpp:128]: This should be `wxSTC_TCS_DEFAULT` ...  it is a `#define` from the TorqueScript lexer.

File: code/Torsion/PlatformWin.cpp [hunk 13]
```diff
@@ -17,7 +17,8 @@
    #define new DEBUG_NEW 
 #endif 
 
-extern wxMenu *wxCurrentPopupMenu;
+//TODO ? extern wxMenu *wxCurrentPopupMenu;
+wxMenu *wxCurrentPopupMenu;
```
Comment (tomspilman) [inline code/Torsion/PlatformWin.cpp:21]: This is defined by wxWidgets itself...

https://github.com/wxWidgets/wxWidgets/blob/master/src/msw/frame.cpp#L67

... looks like it is still supported.  You just need to be sure the version of wxWidgets is compiled with wxUSE_MENUS_NATIVE defined.

File: code/Torsion/PreCompiler.cpp (hunk 1 above)
Comment (tomspilman) [inline code/Torsion/PreCompiler.cpp:414]: Note this is more than just a slightly wrong implementation.  It is completely the wrong code... like not even heading in the right direction.  It makes zero sense to the intent of the original code.

The original code was right and just needs to be ported correctly.  Any other change here is a waste of time.

Comment (ghost) [inline code/Torsion/PreCompiler.cpp:414]: Oh wow. I'm sure glad you mentioned that. I was under the impression that some crap I slapped in there to get it to compile might actually WORK! Phew. good thing you let me know. That was a close one.

File: code/Torsion/ScriptCtrl.cpp [hunk 14]
```diff
@@ -176,7 +178,7 @@ bool ScriptCtrl::LoadFile( const wxString& filename )
    if ( TextBuffer )
    {
       TextBuffer->Lock();
-      GetTextRangeRaw( TextBuffer->GetWriteBuf( GetTextLength() ), 0, -1 );
+      GetTextRangeRaw( wxAtoi(TextBuffer->GetWriteBuf( GetTextLength() )), 0 );
```
Comment (tomspilman) [inline code/Torsion/ScriptCtrl.cpp:181]: This is wrong...  there was no intent to convert the text to a number like you are doing here.

This function was something i hacked into STC...

https://github.com/SickheadGames/Torsion/blob/master/code/wxWidgets/contrib/src/stc/stc.cpp#L3323

It was mainly an optimization to avoid an extra copy.  The right fix for now would be to call `GetTextRangeRaw(0, -1)` then copy the resulting buffer into the buffer returned by `TextBuffer->GetWriteBuf( GetTextLength() )`.

We should leave a `TODO` comment noting the original optimization.

File: code/Torsion/ScriptDoc.cpp [hunk 15]
```diff
@@ -55,7 +55,7 @@ void ScriptDoc::OnAutoCompHint( bool enabled )
       {
          int length = view->GetCtrl()->GetTextLength();
          wxChar* buffer = m_TextBuffer->GetWriteBuf( length );
-         view->GetCtrl()->GetTextRangeRaw( buffer, 0, -1 );
+         //TODO ? view->GetCtrl()->GetTextRangeRaw( buffer, 0 );
```
Comment (tomspilman) [inline code/Torsion/ScriptDoc.cpp:58]: This is the same case as what i wrote for `GetTextRangeRaw` above.  Call the original function then copy the data back into `buffer`.

File: code/Torsion/ScriptFrame.cpp [hunk 16]
```diff
@@ -95,7 +95,8 @@ void ScriptFrame::AddScriptPage( ScriptView* scriptView, wxWindow* ctrl )
    wxDocument* doc = scriptView->GetDocument();
    wxASSERT( doc );
    wxString title;
-   doc->GetPrintableName(title);
+   //TODO ? doc->GetPrintableName(title);
+   title = doc->GetTitle();
```
Comment (tomspilman) [inline code/Torsion/ScriptFrame.cpp:99]: Seems like a good assumption for now.  We will need to research what `GetPrintableName` did 10 years ago in wxWidgets.

File: code/Torsion/ScriptScanner.cpp [hunk 17]
```diff
@@ -124,10 +124,10 @@ bool ScriptScanner::Open( const wxString& path )
    size_t len = (size_t)file.Length();
    if ( len > 0 )
    {
-      wxChar* buff = m_Memory.GetWriteBuf(len);
+	   wxChar* buff = new wxChar(m_Memory.GetChar(len));
```
Comment (tomspilman) [inline code/Torsion/ScriptScanner.cpp:127]: As noted in the other locations... restore the original code.

File: code/Torsion/ScriptView.cpp [hunk 18]
```diff
@@ -204,7 +204,8 @@ bool ScriptView::ProcessEvent( wxEvent& event )
             event.GetEventType() == EVT_COMMAND_BREAKPOINT ) )
    {
       wxPropagationDisabler disable( event );
-      if ( m_ScriptCtrl->ProcessEvent( event ) )
+      //TODO ? if ( m_ScriptCtrl->ProcessEvent( event ) )
+	  if(ProcessEvent(event))
```
Comment (tomspilman) [inline code/Torsion/ScriptView.cpp:208]: This is bad...  it would make the event be processed by the view a second time causing it to re-enter this function in an infinite loop.  I suspect we want `m_ScriptCtrl->ProcessWindowEvent(event)` here too.

File: code/Torsion/ScriptView.cpp [hunk 19]
```diff
@@ -543,7 +545,8 @@ void ScriptView::OnBreakpointEvent( BreakpointEvent& event )
    // TODO: Should we just make BreakpointEvent a wxCommandEvent
    // which would then be forwarded by ScriptView::ProcessEvent?
    wxASSERT( m_ScriptCtrl );
-   m_ScriptCtrl->ProcessEvent( event );
+   //TODO ?   m_ScriptCtrl->ProcessEvent( event );
+   ProcessEvent(event);
```
Comment (tomspilman) [inline code/Torsion/ScriptView.cpp:549]: Same as above.

File: code/Torsion/TorsionApp.cpp [hunk 20]
```diff
@@ -392,7 +392,7 @@ void TorsionApp::RegisterScriptExts()
          // The extension does not exist... so we create 
          // it and own it.
          key.Create();
-         key.SetValue( NULL, "TorsionTorqueScript" );
+         key.SetValue( wxEmptyString, "TorsionTorqueScript" );
```
Comment (tomspilman) [inline code/Torsion/TorsionApp.cpp:395]: Why was this needed?

File: code/Torsion/UpdateChecker.cpp [hunk 21]
```diff
@@ -59,9 +59,9 @@ bool UpdateChecker::CheckAvailable( wxString* version )
    {
       int startLen = data.Len();
       int newLen = startLen + 1024;
-      wxChar* buf = data.GetWriteBuf( newLen ) + startLen;
+      wxChar* buf = new wxChar(data.GetChar( newLen + startLen));
       client.Read( buf, newLen - startLen );
-      data.UngetWriteBuf( startLen + client.LastCount() );
+      //data.UngetWriteBuf( startLen + client.LastCount() );
```
Comment (tomspilman) [inline code/Torsion/UpdateChecker.cpp:64]: Again wrong and needs to be reverted.

File: code/Torsion/WatchCtrl.cpp [hunk 22]
```diff
@@ -496,7 +496,7 @@ void ListCtrl::OnChar( wxKeyEvent& Event )
             break;
 
 		case WXK_PAGEUP:
-		case WXK_PRIOR:
+		case WXK_MEDIA_PREV_TRACK:
```
Comment (tomspilman) [inline code/Torsion/WatchCtrl.cpp:499]: This is wrong... we are not looking for media track events here.

From looking at the wxWidgets changes it seems like `WXK_PRIOR` and `WXK_NEXT` the are not needed anymore:

https://github.com/wxWidgets/wxWidgets/blob/master/docs/changes_30.txt#L261

That changes doc is pretty good.  I suggest reading thru it when you find a missing or changed function to help identify the right fix.

File: code/Torsion/shNotebook.h (hunk 2 above)
Comment (tomspilman) [inline code/Torsion/shNotebook.h:18]: Yeah... these exist in wxWidgets today:

https://github.com/wxWidgets/wxWidgets/blob/master/include/wx/notebook.h

... looks like a missing include is all.

File: code/Torsion/tsComboBox.cpp [hunk 23]
```diff
@@ -8,6 +8,6 @@
 
 void tsComboBox::EnableReselection()
 {
-   m_selectionOld = -1;
+   //m_selectionOld = -1;
```
Comment (tomspilman) [inline code/Torsion/tsComboBox.cpp:11]: I'm not sure what the fix is for this one... we will have to look more closely at the old wxWidgets code and the new code and figure it out.

File: code/Torsion/AutoCompText.cpp (hunk 3 above)
Comment (tomspilman) [inline code/Torsion/AutoCompText.cpp:71]: > I am going to attempt to move to unicode as well

I recommend not doing that for this PR.  Simply move to wxWidgets ansii.

The move to unicode is a totally separate issue and we will deal with it, but just not now.

File: code/Torsion/GotoDefDialog.cpp (hunk 7 above)
Comment (tomspilman) [inline code/Torsion/GotoDefDialog.cpp:241]: I'm doing it to help... we all want to see this working.  Having a pull request invites people to review your code...  if you don't want code review you should close this PR and reopen it when you do.

File: code/Torsion/PreCompiler.cpp (hunk 1 above)
Comment (tomspilman) [inline code/Torsion/PreCompiler.cpp:414]: @blackwc - Look... we are trying to help you.  That is all.

The obvious way to make things just compile is commenting it out.  Replacing it with random code that does something completely different from the original makes other coders think you have no clue what you are doing.  It looks like inexperience...  not a work around.

I'll make note of your quirk for replacing working code with broken code and wait for you to finish.

However cut out the attitude.  No one here is your enemy or attacking you.

Comment (ghost) [inline code/Torsion/PreCompiler.cpp:414]: If I had commented out a lot of the stuff then I would have just had to go back in there and set variables etc. so I just slapped whatever in there because I wanted to get all the includes correctly placed and have it compile.

The attitude is because it was already explained in this conversation thread what I did and why I did it. There was absolutely no reason for you to spew your opinion about it as it was essentially the same as the other critique which I had answered.

Comment (tomspilman) [inline code/Torsion/PreCompiler.cpp:414]: I apologize if me warning you that your code is wrong set you off.  I've never seen any coder replace code with nonsense and even with your warning I wanted to be sure you knew you're heading in the wrong direction.

This is my project.  I will write my opinion on pull requests.  I won't be right at times and misunderstandings occur when communicating here.  I hope you understand that moving forward and don't take things I write as an insult.

**Comment by ghost** [comment]
Torsion now gives me a bad taste in my mouth. I'm closing this PR but leaving the branch up.

**Comment by tomspilman** [comment]
@blackwc - That is fine.  Good luck with your projects.

File: code/Torsion/PreCompiler.cpp (hunk 1 above)
Comment (ghost) [inline code/Torsion/PreCompiler.cpp:414]: What set me off is that there was no reason to pile on.

What I did was no different than setting a bunch of variables to zero except it was quicker and allowed me to efficiently get all the includes in order so it would compile. I marked all instances with "TODO ?" to signify where I put the placeholder code.

What I did was efficient for the task I set out to do which was get every thing in order so it would at least compile and also leave notations so people would know where the nonfunctional placeholder code is.

It was my fault for PRing a branch too soon in development and allowing over eager first-time repo-managers a chance to dig around in a very rough draft.

I'm done with it but I'm leaving the branch up.

Comment (tomspilman) [inline code/Torsion/PreCompiler.cpp:414]: I agree I am over eager, but I may have more repo management experience then you think. :)

Comment (ghost) [inline code/Torsion/PreCompiler.cpp:414]: than*

Comment (tomspilman) [inline code/Torsion/PreCompiler.cpp:414]: Hah... thanks.  My fingers do most of my spelling for me. :)
